### PR TITLE
Flesh out the verify-exercises command

### PR DESCRIPTION
**Pull request by ErikSchierboom**
This PR aims to have new tracks start out with a more full-fleshed `verify-exercises` script.
This should help them get CI working more easily.

@glennj @IsaacG would you mind reviewing this PR? You're both bash experts and I'm decidely not :)

### CODE REVIEW

File: bin/verify-exercises [hunk 1]
```diff
@@ -1,40 +1,79 @@
...
+set -eo pipefail
 
-# Verify the Concept Exercises
-for concept_exercise_dir in ./exercises/concept/${slug}/; do
-    if [ -d $concept_exercise_dir ]; then
-        echo "Checking $(basename "${concept_exercise_dir}") exercise..."
-        # TODO: run command to verify that the exemplar solution passes the tests
-    fi
-done
+required_tool() {
```
Comment (glennj) [inline bin/verify-exercises:15]:
```suggestion
die() { echo "$*" >&2; exit 1; }

required_tool() {
```
Missing `die` function.

File: bin/verify-exercises [hunk 2]
```diff
@@ -1,40 +1,79 @@
...
+    command -v "${1}" >/dev/null 2>&1 ||
+        die "${1} is required but not installed. Please install it and make sure it's in your PATH."
+}
+
+required_tool jq
+
+copy_example_or_examplar_to_solution() {
+    jq -c '[.files.solution, .files.exemplar // .files.example] | transpose | map({src: .[1], dst: .[0]}) | .[]' .meta/config.json | while read -r src_and_dst; do
```
Comment (glennj) [inline bin/verify-exercises:23]:
```suggestion
    jq -c '[.files.solution, .files.exemplar // .files.example] | transpose | map({src: .[1], dst: .[0]}) | .[]' .meta/config.json \
    | while read -r src_and_dst; do
```
Just adding a newline for readability.

Are there exercises where the number of solution files != number of example files?

File: bin/verify-exercises [hunk 3]
```diff
@@ -1,40 +1,79 @@
...
+    local slug
+    local tmp_dir
+
+    dir=$(realpath "${1}")
+    slug=$(basename "${dir}")
+    tmp_dir=$(mktemp -d -t "exercism-verify-${slug}-XXXXX")
+
+    echo "Verifying ${slug} exercise..."
+
+    (
```
Comment (glennj) [inline bin/verify-exercises:62]:
```suggestion
    (
        trap 'rm -rf "$tmp_dir"' EXIT    # remove tempdir when subshell ends
```

File: bin/verify-exercises [hunk 4]
```diff
@@ -1,40 +1,79 @@
...
-for practice_exercise_dir in ./exercises/practice/${slug}/; do
-    if [ -d $practice_exercise_dir ]; then
-        echo "Checking $(basename "${practice_exercise_dir}") exercise..."
-        # TODO: run command to verify that the example solution passes the tests
+shopt -s nullglob
+for exercise_dir in ./exercises/{concept,practice}/${exercise_slug}/; do
+    if [ -d "${exercise_dir}" ]; then
+        verify_exercise "${exercise_dir}"
     fi
 done
```
Comment (glennj) [inline bin/verify-exercises:79]: A bit of feedback to the user
```suggestion
count=0
for exercise_dir in ./exercises/{concept,practice}/${exercise_slug}/; do
    if [[ -d "${exercise_dir}" ]]; then
        verify_exercise "${exercise_dir}"
        ((++count))
    fi
done
((count > 0)) || die 'no matching exercises found!'
```

**Review by glennj — CHANGES_REQUESTED** [review]

File: bin/verify-exercises (hunk 2 above)
Comment (ErikSchierboom) [reply bin/verify-exercises]: No, they should always match

**Review by glennj — APPROVED** [review]
A nice improvement with all the new comments.

File: bin/verify-exercises [hunk 5]
```diff
@@ -1,40 +1,86 @@
...
-# before running the tests.
-#
-# The path to the solution/example/exemplar files can be found in the exercise's
-# .meta/config.json file, or possibly inferred from the exercise's directory name.
+# Verify that each exercise's example/exemplar solution passes the tests.
+# You can either verify all exercises or a single exercise.
 
 # Example: verify all exercises
-# ./bin/verify-exercises
+# bin/verify-exercises
```
Comment (IsaacG) [inline bin/verify-exercises:8]: Doesn't this assume `.` is in `PATH`?

File: bin/verify-exercises [hunk 6]
```diff
@@ -1,40 +1,86 @@
...
+    command -v "${1}" >/dev/null 2>&1 ||
+        die "${1} is required but not installed. Please install it and make sure it's in your PATH."
+}
+
+required_tool jq
+
+copy_example_or_examplar_to_solution() {
+    jq -c '[.files.solution, .files.exemplar // .files.example] | transpose | map({src: .[1], dst: .[0]}) | .[]' .meta/config.json \
+    | while read -r src_and_dst; do
+        cp "$(echo "${src_and_dst}" | jq -r '.src')" "$(echo "${src_and_dst}" | jq -r '.dst')"
```
Comment (IsaacG) [inline bin/verify-exercises:27]:
```suggestion
        cp "$(jq -r '.src' <<< "${src_and_dst}")" "$(jq -r '.dst' <<< "${src_and_dst}")"
```

**Review by IsaacG — APPROVED** [review]

File: bin/verify-exercises (hunk 5 above)
Comment (glennj) [reply bin/verify-exercises:8]: No. It's a relative path, no PATH lookup.

**Comment by ErikSchierboom** [comment]
Thanks both!

**Pull request merged**